java-checks/src/main/java/org/sonar/java/checks/security/CryptographicKeySizeCheck.java

```diff
@@ -0,0 +1,135 @@
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition(KEY_GENERATOR).name(GET_INSTANCE_METHOD).addParameter(STRING),
+      MethodMatcher.create().typeDefinition(KEY_PAIR_GENERATOR).name(GET_INSTANCE_METHOD).addParameter(STRING),
+      MethodMatcher.create().typeDefinition(KEY_PAIR_GENERATOR).name("initialize").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(KEY_GENERATOR).name("init").withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (isGetInstanceMethod(mit)) {
+      storeSymbolOfAssignedInstance(mit);
+    } else if (isValidKeyGenSymbolOfInitializeMethod(mit)) {
+      checkAlgorithmParameterToReport(mit.arguments().get(0)).ifPresent(reportString -> reportIssue(mit, reportString));
+    }
+  }
+
+  private static boolean isValidKeyGenSymbolOfInitializeMethod(MethodInvocationTree mit) {
+    boolean isValidKeyGenSymbol = false;
+    if (mit.methodSelect().is(Tree.Kind.MEMBER_SELECT) && getInstanceInfo.keyGenMethodSymbol != null) {
+      ExpressionTree expr = ((MemberSelectExpressionTree) mit.methodSelect()).expression();
+      if (expr.is(Tree.Kind.IDENTIFIER)) {
+        isValidKeyGenSymbol = getInstanceInfo.keyGenMethodSymbol.equals(((IdentifierTree) expr).symbol());
+      } else if (expr.is(Tree.Kind.MEMBER_SELECT)) {
+        isValidKeyGenSymbol = getInstanceInfo.keyGenMethodSymbol.equals(((MemberSelectExpressionTree) expr).identifier().symbol());
+      }
+    }
+    return isValidKeyGenSymbol;
+  }
+
+  private static boolean isGetInstanceMethod(MethodInvocationTree mit) {
+    boolean algorithmMatches = false;
+    if (GET_INSTANCE_METHOD.equals(mit.symbol().name())) {
+      ExpressionTree arg = (mit.arguments().get(0));
+      if (!arg.is(Tree.Kind.STRING_LITERAL)) {
+        return false;
+      }
+      String algorithm = LiteralUtils.trimQuotes(((LiteralTree) arg).value());
+      if ("Blowfish".equals(algorithm)) {
+        getInstanceInfo.minKeySizeValue = 128;
+        algorithmMatches = true;
+
+      } else if ("RSA".equals(algorithm)) {
+        getInstanceInfo.minKeySizeValue = 2048;
+        algorithmMatches = true;
+      }
+    }
+    return algorithmMatches;
+  }
+
+  private static Optional<String> checkAlgorithmParameterToReport(ExpressionTree argument) {
+    String reportString = null;
+    if (argument.is(Tree.Kind.INT_LITERAL) && (Integer.parseInt(((LiteralTree) argument).value()) < getInstanceInfo.minKeySizeValue)) {
+      reportString = "Use a key length of at least " + getInstanceInfo.minKeySizeValue + " bits.";
+    }
+    getInstanceInfo.nullifyAttributes();
+    return Optional.ofNullable(reportString);
+  }
+
+  private static void storeSymbolOfAssignedInstance(MethodInvocationTree mit) {
```

Comment:
Why not returning a `Symbol` and rename the method `extractSymbolOfAssignedInstance`? This way you can get rid of the global state inside the `GetInstanceMethodInfo.keyGenMethodSymbol`

Reply:
as discussed offline, the rule needs the class-level state because must be invoked twice to find an issue: first for the getInstance() call, second for the initialization call

Reply:
No, that doesn't require class-level state, it can be handled by instance-level state.